src/overlay/TxFloodManager.cpp

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    auto peers = mApp.getOverlayManager().getRandomAuthenticatedPeers();
+
+    UnorderedMap<Peer::pointer, std::pair<TxDemandVector, std::list<Hash>>>
+        demandMap;
+    bool anyNewDemand = false;
+    do
+    {
+        anyNewDemand = false;
+        for (auto const& peer : peers)
+        {
+            auto& demPair = demandMap[peer];
+            auto& demand = demPair.first;
+            auto& retry = demPair.second;
+            bool addedNewDemand = false;
+
+            while (demand.size() < getMaxDemandSize() && peer->hasAdvert() &&
+                   !addedNewDemand)
+            {
+                auto hashPair = peer->popAdvert();
+                auto txHash = hashPair.first;
+                if (hashPair.second)
+                {
+                    auto delta = now - *(hashPair.second);
+                    om.mAdvertQueueDelay.Update(delta);
+                    peer->getPeerMetrics().mAdvertQueueDelay.Update(delta);
+                }
+                switch (demandStatus(txHash, peer))
+                {
+                case DemandStatus::DEMAND:
+                    demand.push_back(txHash);
+                    if (mDemandHistoryMap.find(txHash) ==
+                        mDemandHistoryMap.end())
+                    {
+                        // We don't have any pending demand record of this tx
+                        // hash.
+                        mPendingDemands.push(txHash);
+                        mDemandHistoryMap[txHash].firstDemanded = now;
+                        CLOG_DEBUG(Overlay, "Demand tx {}, asking peer {}",
+                                   hexAbbrev(txHash), peer->toString());
+                    }
+                    else
+                    {
+                        om.mDemandTimeouts.Mark();
+                        ++(peer->getPeerMetrics().mDemandTimeouts);
+                    }
+                    mDemandHistoryMap[txHash].peers.emplace(peer->getPeerID(),
+                                                            now);
+                    mDemandHistoryMap[txHash].lastDemanded = now;
+                    addedNewDemand = true;
+                    break;
+                case DemandStatus::RETRY_LATER:
+                    retry.push_back(txHash);
+                    break;
+                case DemandStatus::DISCARD:
+                    break;
+                }
+            }
+            anyNewDemand |= addedNewDemand;
+        }
+    } while (anyNewDemand);
+
+    for (auto const& peer : peers)
+    {
+        // We move `demand` here and also pass `retry` as a reference
+        // which gets appended. Don't touch `demand` or `retry` after here.
+        peer->sendTxDemand(std::move(demandMap[peer].first));
+        peer->retryAdvert(demandMap[peer].second);
+    }
+
+    // mPendingDemands and mDemandHistoryMap must always contain exactly the
+    // same tx hashes.
+    releaseAssert(mPendingDemands.size() == mDemandHistoryMap.size());
```

Comment:
This assert does not check what the comment says it checks. Maybe that's OK, but the comment should be updated to reflect that the check is intentionally different.

Reply:
I _think_ this was just meant to be a quick sanity check to avoid extra iterations, I'll make a note of it in the comment